packages/foundry/contracts/hooks/OrderHook.sol

```diff
@@ -0,0 +1,476 @@
+import {ReentrancyGuard} from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import {IERC20} from "@openzeppelin/contracts/interfaces/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IPermit2} from "permit2/src/interfaces/IPermit2.sol";
+
+/// @title A Hook to trade coins in a balancer v3 pool
+/// @author Atiq Ishrak
+/// @notice This hook allows users to place orders in a pool and execute them based on the current price of the pool
+contract OrderHook is BaseHooks, VaultGuard, ReentrancyGuard {
+    using SafeERC20 for IERC20;
+
+    /*/////////////////////////////
+    ///////      ERRORS     ///////
+    /////////////////////////////*/
+
+    error OrderHook__UnauthorizedRouter();
+    error OrderHook__OrderAmountInMustBeMoreThanZero();
+    error OrderHook__InsufficientAllowance();
+    error OrderHook__InsufficientBalance();
+    error OrderHook__UnauthorizedCaller();
+    error OrderHook__OrderAlreadyExecutedOrCancelled();
+    error OrderHook__OrderUnableToExecute(
+        bytes32 orderId,
+        address user,
+        uint256 currentTokenInPrice,
+        uint256 triggerPrice,
+        OrderType orderType
+    );
+
+    /*/////////////////////////////
+    ///////      EVENTS     ///////
+    /////////////////////////////*/
+
+    event OrderPlaced(
+        bytes32 indexed orderId,
+        address indexed user,
+        OrderType orderType,
+        uint256 amountIn,
+        uint256 triggerPrice,
+        uint8 slippageTolerance
+    );
+    event OrderExecuted(bytes32 indexed orderId, address indexed user);
+    event OrderCanceled(
+        bytes32 indexed orderId,
+        address indexed user,
+        OrderType orderType
+    );
+    event OrdersProcessed(bytes32[] orderIds);
+    event AfterSwapPrice(
+        address indexed pool,
+        address indexed tokenIn,
+        address indexed tokenOut,
+        uint256 currentTokenInPrice,
+        uint256 currentTokenOutPrice
+    );
+
+    /*/////////////////////////////
+    /////  ENUMS, STRUCTS    //////
+    /////////////////////////////*/
+
+    enum OrderType {
+        STOP_LOSS,
+        BUY_STOP,
+        BUY_LIMIT,
+        TAKE_PROFIT
+    }
+
+    enum OrderStatus {
+        OPEN,
+        EXECUTED,
+        CANCELLED
+    }
+
+    struct Order {
+        bytes32 orderId;
+        address trader;
+        OrderType orderType;
+        OrderStatus orderStatus;
+        uint256 amountIn;
+        uint256 triggerPrice;
+        address tokenIn;
+        address tokenOut;
+        address pool;
+        uint8 slippageTolerance;
+    }
+
+    /*/////////////////////////////
+    ///////    VARIABLES    ///////
+    /////////////////////////////*/
+
+    // only calls from a trusted routers are allowed to call this hook, because the hook relies on the getSender
+    // implementation to work properly
+    address private immutable i_trustedRouter;
```

Comment:
Does it? I don't see any calls to `i_trustedRouter.getSender()`.

Reply:
I think it's fine to store a reference if you need it to swap tokens, but I don't see it using the `getSender` functionality.

Reply:
You're correct. In the current implementation, there isn't any use of `i_trustedRouter.getSender()`. The purpose of `i_trustedRouter` is to verify that the user initiated the swap through the router contract. The documentation was misleading in this regard. It should be updated as follows:
```diff
-   // only calls from a trusted routers are allowed to call this hook, because the hook relies on the getSender
-   // implementation to work properly
+   // only calls from a trusted routers are allowed to call this hook
    address private immutable i_trustedRouter;
```